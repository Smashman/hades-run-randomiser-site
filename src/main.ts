import { Run, RunOptions } from './run';

const runOptions: RunOptions = {
    weaponOptions: {
        enabled: true,
        randomAspect: true,
    },
    keepsakeOptions: {
        enabled: true,
        ignoreMaxLevel: true,
        changeBetweenBiomes: true,
    },
    companionOptions: {
        enabled: true,
        ignoreMaxCodex: true,
    },
    mirrorOptions: {
        enabled: true,
    },
    pactOptions: {
        enabled: true,
        heatRange: {
            min: 4,
            max: 12,
        },
    }
};

const run = Run.generateRandomRun(runOptions);

console.log('run', run);
console.log('run.weapon', run.weapon);
console.log('run.weaponAspect', run.weaponAspect);
console.log('run.keepsake', run.keepsake);
console.log('run.companion', run.companion);
console.log('run.mirror', run.mirror);
console.log('run.pact', run.pact);