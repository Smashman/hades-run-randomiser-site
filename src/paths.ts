export const imgPath = 'public/img';
export const unknownIconPath = `${imgPath}/misc/unknown_icon.png`;