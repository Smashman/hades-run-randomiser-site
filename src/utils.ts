export const getRandomItemFromArray = <T>(array: T[]) => array[Math.floor(Math.random() * array.length)];
export const fillArray = <T>(length: number, fillValue: T) => Array.from<T>({length: length}).fill(fillValue);

export interface Options {
    enabled?: boolean;
}

export abstract class Item {
    isUnlocked: boolean = false;

    unlock() {
        this.isUnlocked = true;
    }

    lock() {
        this.isUnlocked = false;
    }
}

export abstract class ItemList<T extends Item> {
    items: T[] = [];

    addItem(item: T) {
        this.items = [...this.items, item];
        return this;
    }

    addItems(items: T[]) {
        items.forEach(item => this.addItem(item));
        return this;
    }

    get unlocked() {
        return this.items.filter(item => item.isUnlocked);
    }

    unlockAll() {
        this.items.forEach(item => item.unlock());
    }

    getRandom(): T {
        return getRandomItemFromArray(this.unlocked) ?? null;
    }
}

export class Level {
    constructor(protected _value: number, readonly min: number, readonly max: number) {}

    set value(setTo: number) {
        if (setTo >= this.max) {
            this._value = this.max;
        }
        else if (setTo <= this.min) {
            this._value = this.min;
        }
        else {
            this._value = setTo;
        }
    }

    get value() {
        return this._value;
    }

    isMinLevel(): boolean {
        return this.value <= this.min;
    }

    isMaxLevel(): boolean {
        return this.value >= this.max;
    }

    setMaxLevel(): void {
        this.value = this.max;
    }

    addValue(change: number): number {
        this.value = this.value + change;
        return this.value;
    }

    increment(): number {
        return this.addValue(1);
    }

    decrement(): number {
        return this.addValue(-1);
    }
}