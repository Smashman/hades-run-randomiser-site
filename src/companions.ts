import { getRandomItemFromArray, Item, Options } from './utils';
import { ItemList } from './utils';

export class Companion extends Item {
    level: number = 1;
    isMaxCodex: boolean = false;
    constructor(public name: string, public giver: string) {
        super();
    }

    static maxLevel: number = 5;

    get isMaxLevel(): boolean {
        return this.level === Companion.maxLevel;
    }
}

export class Companions extends ItemList<Companion> {
    getRandomCompanion(companionOptions?: CompanionOptions): Companion | null {
        const companions = (companionOptions?.ignoreMaxCodex && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(companion => !companion.isMaxCodex)] : [...this.unlocked];
        return getRandomItemFromArray(companions) ?? null;
    }

    maxCodexAll() {
        this.items.forEach(companion => companion.isMaxCodex = true);
    }

    get isAllMaxCodex() {
        return this.items.every(companion => companion.isMaxCodex);
    }

    maxLevelAll() {
        this.items.forEach(companion => companion.level = Companion.maxLevel);
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(companion => companion.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(companion => companion.isMaxLevel);
    }
}

export interface CompanionOptions extends Options {
    ignoreMaxCodex?: boolean;
}