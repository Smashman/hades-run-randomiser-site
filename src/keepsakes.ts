import { getRandomItemFromArray, Options, ItemList, Item, Level } from './utils';

export interface KeepsakeOptions extends Options {
    ignoreMaxLevel?: boolean;
    changeBetweenBiomes?: boolean;
}

export class Keepsakes extends ItemList<Keepsake> {
    getRandomKeepsakes(keepsakeOptions?: KeepsakeOptions): KeepsakeConfiguration {
        const keepsakes = (keepsakeOptions?.ignoreMaxLevel && !this.isAllUnlockedMaxLevel) ? [...this.unlocked.filter(keepsake => !keepsake.isMaxLevel)] : [...this.unlocked];

        if (keepsakes.length === 0) {
            return null;
        }

        let keepsakesChosen: Keepsake[] = [getRandomItemFromArray(keepsakes)];

        if (keepsakeOptions?.changeBetweenBiomes) {
            for (let i = 1; i <= 3; i++) {
                keepsakesChosen = [...keepsakesChosen, getRandomItemFromArray(keepsakes.filter(keepsake => !keepsakesChosen.includes(keepsake))) ?? null];
            }
        }

        return keepsakesChosen as KeepsakeConfiguration;
    }

    maxLevelAll() {
        this.items.forEach(keepsake => keepsake.level.setMaxLevel());
    }

    get isAllUnlockedMaxLevel() {
        return this.unlocked.every(keepsake => keepsake.isMaxLevel);
    }

    get isAllMaxLevel() {
        return this.items.every(keepsake => keepsake.isMaxLevel);
    }
}

export class Keepsake extends Item {
    level: Level = new Level(1, Keepsake.minLevel, Keepsake.maxLevel);
    constructor(public name: string, public giver: string, public icon: string, public isHidden: boolean = false) {
        super();
    }

    static minLevel = 1;
    static maxLevel = 3;

    get isMaxLevel(): boolean {
        return this.level.isMaxLevel();
    }
}

export type KeepsakeConfiguration = [Keepsake] | [Keepsake, Keepsake, Keepsake, Keepsake] | null;