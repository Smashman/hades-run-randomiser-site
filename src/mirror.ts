import { getRandomItemFromArray, Item, Options, StorableItemData } from './utils';

export interface MirrorOptions extends Options {
};

export const defaultMirrorOptions: MirrorOptions = {
    enabled: true,
};

export class Mirror {
    groups: MirrorGroup[] = [];

    addGroup(sets: [[MirrorTalent, MirrorTalent], [MirrorTalent, MirrorTalent]], isUnlocked: boolean = false) {
        const mirrorTalentSets = <[MirrorTalentSet, MirrorTalentSet]>sets.map(set => new MirrorTalentSet(set));
        this.groups = [...this.groups, new MirrorGroup(mirrorTalentSets, isUnlocked)];
        return this;
    }

    get unlockedGroups() {
        return this.groups.filter(group => group.isUnlocked);
    }

    unlockAll() {
        this.groups.forEach(group => group.unlock());
    }

    getRandom(): MirrorConfiguration {
        return new MirrorConfiguration(this.unlockedGroups.map(group => group.getRandom()).reduce((acc, val) => [...acc, ...val], <MirrorTalent[]>[]));
    }

    toStorableData() {
        return {
            groups: this.groups.map(group => group.toStorableData())
        }
    }
}

export class MirrorGroup extends Item{
    constructor(public sets: [MirrorTalentSet, MirrorTalentSet], public isUnlocked: boolean = false) {
        super();
    }

    getRandom() {
        return <[MirrorTalent, MirrorTalent]>this.sets.map(set => set.getRandom());
    }
}

export class MirrorTalentSet {
    constructor(public talents: [MirrorTalent, MirrorTalent]) {
        talents[0].colour = "red";
        talents[1].colour = "green";
    }

    getRandom(): MirrorTalent {
        return getRandomItemFromArray(this.talents);
    }
}

export class MirrorTalent {
    rankCount: number;
    colour: "red" | "green";
    totalCost: number;

    constructor(public name: string, public rankCosts: number[]) {
        this.rankCount = rankCosts.length;
        this.totalCost = rankCosts.reduce((totalCost, cost) => totalCost + cost);
    }
}

export class MirrorConfiguration {
    totalDarkness: number;
    constructor(public talents: MirrorTalent[]){
        this.totalDarkness = talents.reduce((totalCost, talent) => totalCost + talent.totalCost, 0);
    };
}