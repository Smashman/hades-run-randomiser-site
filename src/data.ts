import { Weapon, WeaponAspect, Weapons } from './weapons';
import { Keepsake, Keepsakes } from './keepsakes';
import { Companion, Companions } from './companions';
import { Mirror, MirrorTalent } from './mirror';
import { Pact, PactCondition } from './pact';
import { fillArray } from './utils';
import * as React from 'react';
import * as keepsakeIcons from './img/keepsake';
import * as weaponIcons from './img/weapon';

export const weapons = new Weapons().addItems([
    new Weapon('Stygius', 'Blade of the Underworld', 'sword', true)
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.sword.zagreus, false, true),
        new WeaponAspect('Nemesis', weaponIcons.sword.nemesis),
        new WeaponAspect('Poseidon', weaponIcons.sword.poseidon),
        new WeaponAspect('Arthur', weaponIcons.sword.arthur, true),
    ]),
    new Weapon('Varatha', 'Eternal Spear', 'spear')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.spear.zagreus, false, true),
        new WeaponAspect('Achilles', weaponIcons.spear.achilles),
        new WeaponAspect('Hades', weaponIcons.spear.hades),
        new WeaponAspect('Guan Yu', weaponIcons.spear.guanYu, true),
    ]),
    new Weapon('Aegis', 'Shield of Chaos', 'shield')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.shield.zagreus, false, true),
        new WeaponAspect('Chaos', weaponIcons.shield.chaos),
        new WeaponAspect('Zeus', weaponIcons.shield.zeus),
        new WeaponAspect('Beowulf', weaponIcons.shield.beowulf, true),
    ]),
    new Weapon('Coronacht', 'Heart-Seeking Bow', 'bow')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.bow.zagreus, false, true),
        new WeaponAspect('Chiron', weaponIcons.bow.chiron),
        new WeaponAspect('Hera', weaponIcons.bow.hera),
        new WeaponAspect('Rama', weaponIcons.bow.rama, true),
    ]),
    new Weapon('Malphon', 'Twin Fists', 'fists')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.fists.zagreus, false, true),
        new WeaponAspect('Talos', weaponIcons.fists.talos),
        new WeaponAspect('Demeter', weaponIcons.fists.demeter),
        new WeaponAspect('Gilgamesh', weaponIcons.fists.gilgamesh, true),
    ]),
    new Weapon('Exagryph', 'Adamant Rail', 'gun')
    .addAspects([
        new WeaponAspect('Zagreus', weaponIcons.gun.zagreus, false, true),
        new WeaponAspect('Eris', weaponIcons.gun.eris),
        new WeaponAspect('Hestia', weaponIcons.gun.hestia),
        new WeaponAspect('Lucifer', weaponIcons.gun.lucifer, true),
    ]),
]);

export const keepsakes = new Keepsakes().addItems([
    new Keepsake('Old Spiked Collar', 'Cerberus', keepsakeIcons.oldSpikedCollar),
    new Keepsake('Myrmidon Bracer', 'Achilles', keepsakeIcons.myrmidonBracer),
    new Keepsake('Black Shawl', 'Nyx', keepsakeIcons.blackShawl),
    new Keepsake('Pierced Butterfly', 'Thanatos', keepsakeIcons.piercedButterfly),
    new Keepsake('Bone Hourglass', 'Charon', keepsakeIcons.boneHourglass),
    new Keepsake('Chthonic Coin Purse', 'Hypnos', keepsakeIcons.chthonicCoinPurse),
    new Keepsake('Skull Earring', 'Megaera', keepsakeIcons.skullEarring),
    new Keepsake('Distant Memory', 'Orpheus', keepsakeIcons.distantMemory),
    new Keepsake('Harpy Feather Duster', 'Dusa', keepsakeIcons.harpyFeatherDuster),
    new Keepsake('Lucky Tooth', 'Skelly', keepsakeIcons.luckyTooth),
    new Keepsake('Thunder Signet', 'Zeus', keepsakeIcons.thunderSignet),
    new Keepsake('Conch Shell', 'Poseidon', keepsakeIcons.conchShell),
    new Keepsake('Owl Pendant', 'Athena', keepsakeIcons.owlPendant),
    new Keepsake('Eternal Rose', 'Aphrodite', keepsakeIcons.eternalRose),
    new Keepsake('Blood-Filled Vial', 'Ares', keepsakeIcons.bloodFilledVial),
    new Keepsake('Adamant Arrowhead', 'Artemis', keepsakeIcons.adamantArrowhead),
    new Keepsake('Overflowing Cup', 'Dionysus', keepsakeIcons.overflowingCup),
    new Keepsake('Lambent Plume', 'Hermes', keepsakeIcons.lambentPlume),
    new Keepsake('Frostbitten Horn', 'Demeter', keepsakeIcons.frostbittenHorn),
    new Keepsake('Cosmic Egg', 'Chaos', keepsakeIcons.cosmicEgg),
    new Keepsake('Shattered Shackle', 'Sisyphus', keepsakeIcons.shatteredShackle),
    new Keepsake('Evergreen Acorn', 'Eurydice', keepsakeIcons.evergreenAcorn),
    new Keepsake('Broken Spearpoint', 'Patroclus', keepsakeIcons.brokenSpearpoint),
    new Keepsake('Pom Blossom', 'Persephone', keepsakeIcons.pomBlossom),
    new Keepsake('Sigil of the Dead', 'Hades', keepsakeIcons.sigilOfTheDead, true),
]);

export const companions = new Companions().addItems([
    new Companion('Battie', 'Megaera'),
    new Companion('Mort', 'Thanatos'),
    new Companion('Rib', 'Skelly'),
    new Companion('Shady', 'Sisyphus'),
    new Companion('Fidi', 'Dusa'),
    new Companion('Antos', 'Achilles'),
]);

export const mirror = new Mirror()
    .addGroup([
        [
            new MirrorTalent('Shadow Presence', [10, 15, 20, 25, 30]),
            new MirrorTalent('Fiery Presence', [10, 15, 20, 25, 30]),
        ], [
            new MirrorTalent('Chthonic Vitality', [10, 20, 40]),
            new MirrorTalent('Dark Regeneration', [30, 60]),
        ]
    ], true)
    .addGroup([
        [
            new MirrorTalent('Death Defiance', [30, 500, 1000]),
            new MirrorTalent('Stubborn Defiance', [600]),
        ],
        [
            new MirrorTalent('Greater Reflex', [50]),
            new MirrorTalent('Ruthless Reflex', [75]),
        ]
    ], true)
    .addGroup([
        [
            new MirrorTalent('Boiling Blood', [10, 30, 50, 70, 90]),
            new MirrorTalent('Abyssal Blood', [20, 40, 60, 80, 100]),
        ],
        [
            new MirrorTalent('Infernal Soul', [20, 80]),
            new MirrorTalent('Stygian Soul', [0, 60, 180]),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Deep Pockets', [30, 35, 40, 45, 50, 55, 60, 65, 70, 75]),
            new MirrorTalent('Golden Touch', [70, 90, 110]),
        ],
        [
            new MirrorTalent('Thick Skin', [40, 45, 50, 55, 60, 65, 70, 75, 80, 85]),
            new MirrorTalent('High Confidence', [50, 100, 150, 200, 250]),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Privileged Status', [50, 500]),
            new MirrorTalent('Family Favorite', [50, 150]),
        ],
        [
            new MirrorTalent('Olympian Favor', fillArray(40, 50)),
            new MirrorTalent('Dark Foresight', fillArray(10, 150)),
        ]
    ])
    .addGroup([
        [
            new MirrorTalent('Gods\' Pride', fillArray(20, 100)),
            new MirrorTalent('Gods\' Legacy', fillArray(10, 250)),
        ],
        [
            new MirrorTalent('Fated Authority', [500, 750, 1000, 1250, 1500, 1750, 2000, 2250]),
            new MirrorTalent('Fated Persuasion', [1000, 2000, 3000, 4000]),
        ]
    ]);

export const pact = new Pact()
    .addCondition(new PactCondition('Hard Labor', [1, 1, 1, 1, 1]))
    .addCondition(new PactCondition('Lasting Consequences', [1, 1, 1, 1]))
    .addCondition(new PactCondition('Convenience Fee', [1, 1]))
    .addCondition(new PactCondition('Jury Summons', [1, 1, 1]))
    .addCondition(new PactCondition('Extreme Measures', [1, 2, 3, 4], 3))
    .addCondition(new PactCondition('Calisthenics Program', [1, 1]))
    .addCondition(new PactCondition('Benefits Package', [2, 3]))
    .addCondition(new PactCondition('Middle Management', [2]))
    .addCondition(new PactCondition('Underworld Customs', [2]))
    .addCondition(new PactCondition('Forced Overtime', [3, 3]))
    .addCondition(new PactCondition('Heightened Security', [1]))
    .addCondition(new PactCondition('Routine Inspection', [2, 2, 2, 2]))
    .addCondition(new PactCondition('Damage Control', [1, 1]))
    .addCondition(new PactCondition('Approval Process', [2, 3]))
    .addCondition(new PactCondition('Tight Deadline', [1, 2, 3]));

// weapons.unlockAll();
// keepsakes.unlockAll();
// companions.unlockAll();
// mirror.unlockAll();
// pact.unlockAll();

// weapons.maxLevelAll();
// keepsakes.maxLevelAll();
// companions.maxLevelAll();
// companions.maxCodexAll();

interface Data {
    weapons: Weapons;
    keepsakes: Keepsakes;
    companions: Companions;
    mirror: Mirror;
    pact: Pact;
}

export const defaultData: Data = {
    weapons,
    keepsakes,
    companions,
    mirror,
    pact,
};

export const DataContext = React.createContext<[Data, React.Dispatch<React.SetStateAction<Data>>]>([defaultData, {} as React.Dispatch<React.SetStateAction<Data>>]);

//@ts-ignore
// window.data = defaultData;