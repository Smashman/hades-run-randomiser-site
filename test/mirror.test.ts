import { MirrorTalent, MirrorTalentSet, MirrorGroup, Mirror, MirrorConfiguration } from '../src/mirror';
import * as utils from '../src/utils';
import { getRandomItemFromArrayMock } from './helpers';

describe('MirrorTalent class', () => {
    it('should create expected MirrorTalent class instance', () => {
        const mirrorTalent = new MirrorTalent('Run Fast', [1, 2, 3]);

        expect(mirrorTalent).toMatchSnapshot();
        expect(mirrorTalent.colour).toBe(undefined);
    });
});

describe('MirrorTalentSet class', () => {
    it('should create expected MirrorTalentSet class instance', () => {
        const mirrorTalentSet = new MirrorTalentSet([new MirrorTalent('Run Fast', [1, 2, 3]), new MirrorTalent('Jump High', [3, 3, 3])]);

        expect(mirrorTalentSet).toMatchSnapshot();
    });

    it('should set first talent as red and second talent as green', () => {
        const mirrorTalent1 = new MirrorTalent('Run Fast', [1, 2, 3]);
        const mirrorTalent2 = new MirrorTalent('Jump High', [3, 3, 3]);
        new MirrorTalentSet([mirrorTalent1, mirrorTalent2]);

        expect(mirrorTalent1.colour).toBe('red');
        expect(mirrorTalent2.colour).toBe('green');
    });

    it('should return expected talent when calling getRandom', () => {
        const getRandomItemFromArraySpy = jest.spyOn(utils, 'getRandomItemFromArray').mockImplementation(getRandomItemFromArrayMock);


        const mirrorTalent1 = new MirrorTalent('Run Fast', [1, 2, 3]);
        const mirrorTalent2 = new MirrorTalent('Jump High', [3, 3, 3]);
        const mirrorTalentSet = new MirrorTalentSet([mirrorTalent1, mirrorTalent2]);

        const randomMirrorTalent = mirrorTalentSet.getRandom();

        expect(getRandomItemFromArraySpy).toBeCalledWith([mirrorTalent1, mirrorTalent2]);
        expect(randomMirrorTalent).toBe(mirrorTalent2);

        getRandomItemFromArraySpy.mockRestore();
    });
});

describe('MirrorGroup class', () => {
    let mirrorTalent1: MirrorTalent;
    let mirrorTalent2: MirrorTalent;
    let mirrorTalent3: MirrorTalent;
    let mirrorTalent4: MirrorTalent;

    let mirrorTalentSet1: MirrorTalentSet;
    let mirrorTalentSet2: MirrorTalentSet;

    beforeEach(() => {
        mirrorTalent1 = new MirrorTalent('Run Fast', [1, 2, 3]);
        mirrorTalent2 = new MirrorTalent('Jump High', [3, 3, 3]);
        mirrorTalent3 = new MirrorTalent('Talk Good', [5, 5, 5, 5]);
        mirrorTalent4 = new MirrorTalent('Smile Nice', [20]);

        mirrorTalentSet1 = new MirrorTalentSet([mirrorTalent1, mirrorTalent2]);
        mirrorTalentSet2 = new MirrorTalentSet([mirrorTalent3, mirrorTalent4]);
    });

    it('should create expected MirrorGroup class instance', () => {
        const mirrorTalentGroup = new MirrorGroup([mirrorTalentSet1, mirrorTalentSet2]);

        expect(mirrorTalentGroup).toMatchSnapshot();
        expect(mirrorTalentGroup.isUnlocked).toBe(false);
    });

    it('should create expected MirrorGroup class instance with isUnlocked set to true', () => {
        const mirrorTalentGroup = new MirrorGroup([mirrorTalentSet1, mirrorTalentSet2], true);

        expect(mirrorTalentGroup.isUnlocked).toBe(true);
    });

    it('should return expected talents when calling getRandom', () => {
        const getRandomItemFromArraySpy = jest.spyOn(utils, 'getRandomItemFromArray').mockImplementation(getRandomItemFromArrayMock);

        const mirrorTalentGroup = new MirrorGroup([mirrorTalentSet1, mirrorTalentSet2]);

        const randomMirrorTalents = mirrorTalentGroup.getRandom();

        expect(getRandomItemFromArraySpy).toBeCalledTimes(2);
        expect(randomMirrorTalents).toStrictEqual([mirrorTalent2, mirrorTalent4]);

        getRandomItemFromArraySpy.mockRestore();
    });
});

describe('MirrorConfiguration class', () => {
    it('should create expected MirrorConfiguration class instance', () => {
        const mirrorConfiguration = new MirrorConfiguration([new MirrorTalent('Run Fast', [1, 2, 3]), new MirrorTalent('Talk Good', [5, 5, 5, 5])]);

        expect(mirrorConfiguration).toMatchSnapshot();
        expect(mirrorConfiguration.totalDarkness).toBe(26);
    });
});

describe('Mirror class', () => {
    let mirrorTalent1: MirrorTalent;
    let mirrorTalent2: MirrorTalent;
    let mirrorTalent3: MirrorTalent;
    let mirrorTalent4: MirrorTalent;
    let mirrorTalent5: MirrorTalent;
    let mirrorTalent6: MirrorTalent;
    let mirrorTalent7: MirrorTalent;
    let mirrorTalent8: MirrorTalent;

    beforeEach(() => {
        mirrorTalent1 = new MirrorTalent('Run Fast', [1, 2, 3]);
        mirrorTalent2 = new MirrorTalent('Jump High', [3, 3, 3]);
        mirrorTalent3 = new MirrorTalent('Talk Good', [5, 5, 5, 5]);
        mirrorTalent4 = new MirrorTalent('Smile Nice', [20]);
        mirrorTalent5 = new MirrorTalent('Live Long', [100, 100, 100, 100]);
        mirrorTalent6 = new MirrorTalent('Dodge Quick', [15, 15]);
        mirrorTalent7 = new MirrorTalent('Eat Well', [250, 250]);
        mirrorTalent8 = new MirrorTalent('Smell Pleasant', [4000]);
    });

    it('should create expected Mirror class instance', () => {
        const mirror = new Mirror();

        expect(mirror).toMatchSnapshot();
        expect(mirror.unlockedGroups).toStrictEqual([]);
    });

    it('should add group with addGroup', () => {
        const mirror = new Mirror().addGroup([
            [mirrorTalent1, mirrorTalent2],
            [mirrorTalent3, mirrorTalent4],
        ]);

        expect(mirror).toMatchSnapshot();
        expect(mirror.groups[0].isUnlocked).toBe(false);
        expect(mirror.unlockedGroups).toHaveLength(0);
    });

    it('should add unlocked group with addGroup with isUnlocked true', () => {
        const mirror = new Mirror().addGroup([
            [mirrorTalent1, mirrorTalent2],
            [mirrorTalent3, mirrorTalent4],
        ], true);

        expect(mirror.groups[0].isUnlocked).toBe(true);
        expect(mirror.unlockedGroups).toHaveLength(1);
    });

    it('should unlock all groups with unlockAll', () => {
        const mirror = new Mirror().addGroup([
            [mirrorTalent1, mirrorTalent2],
            [mirrorTalent3, mirrorTalent4],
        ]).addGroup([
            [mirrorTalent5, mirrorTalent6],
            [mirrorTalent7, mirrorTalent8],
        ]);

        expect(mirror.groups[0].isUnlocked).toBe(false);
        expect(mirror.groups[1].isUnlocked).toBe(false);
        expect(mirror.unlockedGroups).toHaveLength(0);

        mirror.unlockAll();

        expect(mirror.groups[0].isUnlocked).toBe(true);
        expect(mirror.groups[1].isUnlocked).toBe(true);
        expect(mirror.unlockedGroups).toHaveLength(2);
    });

    it('should return expected MirrorConfiguration from getRandom with no unlocked groups', () => {
        const getRandomItemFromArraySpy = jest.spyOn(utils, 'getRandomItemFromArray').mockImplementation(getRandomItemFromArrayMock);

        const mirror = new Mirror().addGroup([
            [mirrorTalent1, mirrorTalent2],
            [mirrorTalent3, mirrorTalent4],
        ]).addGroup([
            [mirrorTalent5, mirrorTalent6],
            [mirrorTalent7, mirrorTalent8],
        ]);

        const randomMirrorConfiguration = mirror.getRandom();

        expect(getRandomItemFromArraySpy).toBeCalledTimes(0);
        expect(randomMirrorConfiguration).toStrictEqual(new MirrorConfiguration([]));

        getRandomItemFromArraySpy.mockRestore();
    });

    it('should return expected MirrorConfiguration from getRandom with unlocked groups', () => {
        const getRandomItemFromArraySpy = jest.spyOn(utils, 'getRandomItemFromArray').mockImplementation(getRandomItemFromArrayMock);

        const mirror = new Mirror().addGroup([
            [mirrorTalent1, mirrorTalent2],
            [mirrorTalent3, mirrorTalent4],
        ]).addGroup([
            [mirrorTalent5, mirrorTalent6],
            [mirrorTalent7, mirrorTalent8],
        ]);

        mirror.unlockAll();

        const randomMirrorConfiguration = mirror.getRandom();

        expect(getRandomItemFromArraySpy).toBeCalledTimes(4);
        expect(randomMirrorConfiguration).toMatchSnapshot();
        expect(randomMirrorConfiguration.talents).toStrictEqual([mirrorTalent2, mirrorTalent4, mirrorTalent6, mirrorTalent8]);

        getRandomItemFromArraySpy.mockRestore();
    });
});