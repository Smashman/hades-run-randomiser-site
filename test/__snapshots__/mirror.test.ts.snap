// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mirror class should add group with addGroup 1`] = `
Mirror {
  "groups": Array [
    MirrorGroup {
      "isUnlocked": false,
      "sets": Array [
        MirrorTalentSet {
          "talents": Array [
            MirrorTalent {
              "colour": "red",
              "name": "Run Fast",
              "rankCosts": Array [
                1,
                2,
                3,
              ],
              "rankCount": 3,
              "totalCost": 6,
            },
            MirrorTalent {
              "colour": "green",
              "name": "Jump High",
              "rankCosts": Array [
                3,
                3,
                3,
              ],
              "rankCount": 3,
              "totalCost": 9,
            },
          ],
        },
        MirrorTalentSet {
          "talents": Array [
            MirrorTalent {
              "colour": "red",
              "name": "Talk Good",
              "rankCosts": Array [
                5,
                5,
                5,
                5,
              ],
              "rankCount": 4,
              "totalCost": 20,
            },
            MirrorTalent {
              "colour": "green",
              "name": "Smile Nice",
              "rankCosts": Array [
                20,
              ],
              "rankCount": 1,
              "totalCost": 20,
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`Mirror class should create expected Mirror class instance 1`] = `
Mirror {
  "groups": Array [],
}
`;

exports[`Mirror class should return expected MirrorConfiguration from getRandom with unlocked groups 1`] = `
MirrorConfiguration {
  "talents": Array [
    MirrorTalent {
      "colour": "green",
      "name": "Jump High",
      "rankCosts": Array [
        3,
        3,
        3,
      ],
      "rankCount": 3,
      "totalCost": 9,
    },
    MirrorTalent {
      "colour": "green",
      "name": "Smile Nice",
      "rankCosts": Array [
        20,
      ],
      "rankCount": 1,
      "totalCost": 20,
    },
    MirrorTalent {
      "colour": "green",
      "name": "Dodge Quick",
      "rankCosts": Array [
        15,
        15,
      ],
      "rankCount": 2,
      "totalCost": 30,
    },
    MirrorTalent {
      "colour": "green",
      "name": "Smell Pleasant",
      "rankCosts": Array [
        4000,
      ],
      "rankCount": 1,
      "totalCost": 4000,
    },
  ],
  "totalDarkness": 4059,
}
`;

exports[`MirrorConfiguration class should create expected MirrorConfiguration class instance 1`] = `
MirrorConfiguration {
  "talents": Array [
    MirrorTalent {
      "name": "Run Fast",
      "rankCosts": Array [
        1,
        2,
        3,
      ],
      "rankCount": 3,
      "totalCost": 6,
    },
    MirrorTalent {
      "name": "Talk Good",
      "rankCosts": Array [
        5,
        5,
        5,
        5,
      ],
      "rankCount": 4,
      "totalCost": 20,
    },
  ],
  "totalDarkness": 26,
}
`;

exports[`MirrorGroup class should create expected MirrorGroup class instance 1`] = `
MirrorGroup {
  "isUnlocked": false,
  "sets": Array [
    MirrorTalentSet {
      "talents": Array [
        MirrorTalent {
          "colour": "red",
          "name": "Run Fast",
          "rankCosts": Array [
            1,
            2,
            3,
          ],
          "rankCount": 3,
          "totalCost": 6,
        },
        MirrorTalent {
          "colour": "green",
          "name": "Jump High",
          "rankCosts": Array [
            3,
            3,
            3,
          ],
          "rankCount": 3,
          "totalCost": 9,
        },
      ],
    },
    MirrorTalentSet {
      "talents": Array [
        MirrorTalent {
          "colour": "red",
          "name": "Talk Good",
          "rankCosts": Array [
            5,
            5,
            5,
            5,
          ],
          "rankCount": 4,
          "totalCost": 20,
        },
        MirrorTalent {
          "colour": "green",
          "name": "Smile Nice",
          "rankCosts": Array [
            20,
          ],
          "rankCount": 1,
          "totalCost": 20,
        },
      ],
    },
  ],
}
`;

exports[`MirrorTalent class should create expected MirrorTalent class instance 1`] = `
MirrorTalent {
  "name": "Run Fast",
  "rankCosts": Array [
    1,
    2,
    3,
  ],
  "rankCount": 3,
  "totalCost": 6,
}
`;

exports[`MirrorTalentSet class should create expected MirrorTalentSet class instance 1`] = `
MirrorTalentSet {
  "talents": Array [
    MirrorTalent {
      "colour": "red",
      "name": "Run Fast",
      "rankCosts": Array [
        1,
        2,
        3,
      ],
      "rankCount": 3,
      "totalCost": 6,
    },
    MirrorTalent {
      "colour": "green",
      "name": "Jump High",
      "rankCosts": Array [
        3,
        3,
        3,
      ],
      "rankCount": 3,
      "totalCost": 9,
    },
  ],
}
`;
